Guacamole integration

calvinmclean opened the pull request:
This PR is required for Guacamole integration. It allows RealVNC Server to allow VncAuth connections rather than just RA2 encrypted connections.

## Related to:
- [Troposphere PR #639](https://github.com/cyverse/troposphere/pull/639) 
- [atmosphere-ansible PR #78](https://github.com/cyverse/atmosphere-ansible/pull/78)
- [Atmosphere PR #381](https://github.com/cyverse/atmosphere/pull/381)
- [Clank PR #167](https://github.com/cyverse/clank/pull/167)

steve-gregory commented on ansible/roles/atmo-realvncserver/tasks/main.yml:
```diff
@@ -123,6 +123,14 @@
+    state: present
+    backrefs: yes
+    regexp: "^SecurityTypes=RA2"
+    line: "SecurityTypes=RA2,VncAuth"
```
When guacamole is disabled we might want to disable this task from running as well..

calvinmclean replied:
What would be the best way to do this? The `GUACAMOLE_ENABLED` variable is on clank in troposphere's group_vars, so I could probably just add the same line: 
`GUACAMOLE_ENABLED: "{{ GUACAMOLE_ENABLED | default(False) }}"`
to the atmosphere_ansible group_vars

c-mart replied:
So far, clank group_vars aren't programmatically related to atmosphere-ansible group_vars.  We do have a new pattern for optional roles/functionality in atmosphere-ansible, see [here](https://github.com/cyverse/atmosphere-ansible#configuring-atmosphere-ansible) and an example playbook [here](https://github.com/cyverse/atmosphere-ansible/blob/master/ansible/playbooks/instance_deploy/30_post_user_install.yml) (but you could also implement the conditional logic inside a role). I would call the variable `SETUP_GUACAMOLE` for consistency with the other "optional functionality" variables.

calvinmclean commented:
I implemented most of the changes from problem 2 in the wiki page:
- Added `SETUP_GUACAMOLE` variable and conditionals
- Port `5905` is used for VNC and RealVNC is supposed to only allow connections from the Guacamole server
- I **did not** add UFW rules to protect the port because UFW isn't usually enabled on the instances

calvinmclean commented:
The recent commits to this branch are intended to make it possible to configure VNC for Guacamole without NoVNC being used. Previously, the Guacamole tasks were integrated into the NoVNC role. In order to make this work, all tasks that are common to NoVNC and Guacamole (basic VNC server setup) have been moved to `atmo-realvncserver`. So `atmo-realvncserver` will setup everything for basic VNC usage and start a VNC server on `5901`. Then, `atmo-novnc` and/or `atmo-guacamole` are run to add additional features.

steve-gregory commented:
@calvinmclean or @c-mart can you address the merge conflicts here?

c-mart commented:
@calvinmclean has been involved much more recently than I have. Calvin, can you take a look and holler if you have questions?

c-mart commented:
Also, overall I approve of the structure of this PR! Hopefully the various permutations (NoVNC only, Guacamole, and both) have been tested.

Additional todo:
- [ ] PRs to environment-specific repos to set new variables as appropriate (SETUP_GUACAMOLE, SETUP_NOVNC)
  - [ ] atmo-prod
  - [ ] atmo-beta
  - [ ] atmo-dev
  - [ ] atmo-local

c-mart commented on ansible/roles/atmo-vnc/tasks/guacamole.yml:
```diff
@@ -0,0 +1,31 @@
+
+- name: guac - Start VNC Server for Guacamole
+  command: >
+    /bin/su {{ ATMOUSERNAME }} -c "/usr/bin/vncserver -config /home/{{ ATMOUSERNAME }}/.vnc/config.guac :5"
```
Is the `:5` what sets port 5905 for the VNC instance?

calvinmclean replied:
@c-mart yes

calvinmclean commented:
Created merge requests for environment-specific variable repos. I will look at the conflicts today and start testing once I have them resolved.

calvinmclean commented:
I have been doing some testing and all looks good. The feature flags `SETUP_GUACAMOLE` and `SETUP_NOVNC` work for Ansible tasks and Troposphere buttons.
I made a small change because the desktop backgrounds did not appear until restarting VNC servers so I just moved the background-setting tasks to run before VNC servers are started.

steve-gregory commented:
Thanks @calvinmclean !
